### ISSUE TITLE
`unnecessary_literal_bound` triggered on trait implementation

### ISSUE THREAD

**Issue by mladedav** [rust-lang/rust-clippy#14085]
### Description

The lint triggers on trait implementations where it forces users to refine the return type instead of following whatever is present in the trait definition.

```rust
#![allow(dead_code)]
#![warn(clippy::unnecessary_literal_bound)]

trait Trait {
    fn foo(&self) -> &str;
}

struct Struct;

impl Trait for Struct {
    fn foo(&self) -> &str {
        "foo"
    }
}

impl Struct {
    fn bar(&self) -> &str {
        "bar"
    }
}

```

I would excpect the lint to only trigger on the `bar` function and not on `foo`.

There's [ongoing discussion](https://github.com/rust-lang/rust/issues/121718) about refining RPITIT and I think this trait should behave similarly to what is done there.

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: x86_64-unknown-linux-gnu
release: 1.84.0
LLVM version: 19.1.5
```

### Additional Labels

@rustbot label +A-lint +L-pedantic +I-false-positive
